Approach 2: WP_Term: Add dynamic property support.

hellofromtonya opened the pull request:
This PR implements Approach 2 to add full support to `WP_Term` for dynamic properties.

### Adds dynamic property support to `WP_Term`.

Removes the `#[AllowDynamicProperties]` attribute from the class.

Adds a full set of magic methods.

Adds a new private property called `$dynamic_properties` to hold all of the dynamic properties that interact with `__get()` and/or `__set()`.

The magic methods work with the new `WP_Term::$dynamic_property`:
* When getting or setting, the dynamic property is added to `WP_Term::$dynamic_property`.
* When unsetting, the dynamic property is removed from `WP_Term::$dynamic_property`.
* When checking if `isset()`, it uses `WP_Term::$dynamic_property`.

### `WP_Term::to_array()`: Adjusts and uses public and dynamic props.

Restores previous behavior of `WP_Term::to_array()` to return an associative array of all the _public and dynamic properties_. It does not include the new private `WP_Term::$dynamic_property` property in the returned array.

Type casting from object to array, e.g. `(array) $term`, is different with the full set of magic methods. It creates an associative array with:

* all of the declared properties (including the new private `WP_Term::$dynamic_properties` property).
* does not include the dynamic properties (as they are now stored in the `WP_Term::$dynamic_properties` property).

Instead of type casting, `WP_Term::to_array()` should be used. Instances in Core are replaced to use it.

## TODO

- [X] Remove the class annotation.
- [X] Add full set of magic methods.
- [x] Populate the array with the dynamic properties in `WP_Term::to_array()`.
- [x] Add unit tests.
- [x] Manually test in local site.
- [ ] Manually test with plugins that (a) [do array type casting](https://www.wpdirectory.net/search/01HWR887XBHDYRN6G0E7N0SVZA) and/or add dynamic properties (such as WooCommerce).

Trac ticket: https://core.trac.wordpress.org/ticket/61890

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

hellofromtonya committed:
Add dynamic property support.

Dynamic properties are stored in a private internal array called 'dynamic_properties',
which are used within the new full set of magic methods.

hellofromtonya committed:
to_array(): Adjust and use for public and dynamic props.

Restores the behavior of WP_Term::to_array() to return an associative array of all the public and dynamic properties.

Type casting from object to array, e.g. `(array) $term`, is different with the full set of magic methods. It creates an associative array with:
* all of the declared properties (including the new private `WP_Term::$dynamic_properties` property).
* does not include the dynamic properties (as they are now stored in the `WP_Term::$dynamic_properties` property).

Instead of type casting, `WP_Term::to_array()` should be used. Instances in Core are replaced to use it.

anton-vlasenko commented:
I've tried to test this PR since it's the preferred solution out of the two, but the unit tests are failing. Could you please take a look, @hellofromtonya?

hellofromtonya commented:
> I've tried to test this PR since it's the preferred solution out of the two, but the unit tests are failing. Could you please take a look, @hellofromtonya?

@anton-vlasenko The tests that are failing are for the `__get()` magic method. I'm still working on that dataset and its test setup. For now, I force pushed a change to skip over that test.

hellofromtonya commented on tests/phpunit/tests/term/wpTerm.php:
```diff
@@ -84,9 +85,296 @@ public function test_get_instance_should_respect_taxonomy_when_term_id_is_found_
+	 * @param mixed  $expected Expected value.
+	 */
+	public function test_should_get_dynamic_property( $name, $args, $expected ) {
+		// This if branch is temporary. Do not commit.
```
I'm working on the additional datasets that trigger this skip. Will force push when ready.

anton-vlasenko commented on src/wp-admin/includes/ajax-actions.php:
```diff
@@ -1165,7 +1165,7 @@ function wp_ajax_add_tag() {
 			'what'         => 'term',
 			'position'     => $level,
-			'supplemental' => (array) $tag,
+			'supplemental' => $tag->to_array(),
```
This change is fine and is necessary to keep adding tags in `wp-admin` working (otherwise JavaScript just silently fails as the spinner keeps spinning on the `edit-tags.php` page despite the tag being added), but it seems that the [Customer Reviews for WooCommerce](https://wordpress.org/plugins/customer-reviews-woocommerce/) plugin may stop working if it doesn't adopt the same change. Source: https://www.wpdirectory.net/search/01J69GH2CS5JB61HFCBF8FGTQS
As noted in https://core.trac.wordpress.org/ticket/61890#comment:13, a developer note will be necessary to inform developers of this plugin, as well as other plugins that convert `WP_Term` objects to arrays, about the upcoming changes.

I initially [avoided](https://core.trac.wordpress.org/ticket/58087#comment:33) this approach because I was uncertain about how it would affect many of the existing plugins, some of which are _very_ popular.

hellofromtonya replied:
I tested the Customer Reviews for WooCommerce plugin with WooCommerce and this patch applied. It worked with no issues.

@anton-vlasenko did it not work for you in your testing?

I'm now thinking the adoption of `$term->to_array()` is not mandatory.

@adrianduffell questions made me realize, [there are 2 distinct use cases for converting the term object into an array](https://core.trac.wordpress.org/ticket/61890#comment:23):
* Use `$term->to_array()` when you need an array that includes all dynamic properties. 
* Use `(array) $term` when you need only the declared properties, i.e. no dynamic properties.

And yes, both of these use cases will be in a dev note for this change.

hellofromtonya replied:
Wait. I see the issue.

Testing it in the Reviews > Review Tagging > Create New Tag. When adding a new review tag, the issue exists where it keeps spinning and does not show the new tag. Upon refresh, the tag is there.

![custom-reviews-for-wc](https://github.com/user-attachments/assets/b77b24b7-67a0-4db0-8836-2921f33d6836)

[Here's the test report](https://core.trac.wordpress.org/ticket/61890#comment:24).

hellofromtonya committed:
Add unit tests.

Splits the tests into 1 method under test per test file/class.

Adds tests for __get(), __set(), __isset(), __unset(), and to_array().

hellofromtonya commented:
@adrianduffell [raised a valid point](https://core.trac.wordpress.org/ticket/61890?replyto=27#comment:27):
>I would see that all PHP language constructs such as (array) type casting would be supported by WordPress, unless forbidden in the coding standards.

This PR (Approach 2) does not meet that expectation.

In order to continue support `(array) $terms` type casting while ensuring it matches the same results as the `WP_Term::to_array()`, this PR is not a viable solution.
